### PR TITLE
Make the flow analysis based unused value/parameters analyzer more conservative

### LINKED ISSUE #32923: Unused variable warning IDE0059 triggered on "discard"
**Version Used**: 
16.0.0 Preview 2

**Steps to Reproduce**:
```csharp
private void test()
{
    ReadOnlySequence<byte> a = default;
    var count = 0;
    foreach (var _ in a) // IDE0059 smudge here on _
        ++count;
}
```

**Expected Behavior**:
VS2017 doesn't show that and using `List<byte>` instead of the `ReadOnlySequence<byte>` doesn't either, so I'm not expecting to see this.

**Actual Behavior**:
IDE0059 is triggered and there's a light bulb is VS. However, trying to click the light bulb that appears when hovering the `_` with the mouse just makes it disappear instead of showing possible fixes. The light bulb at the line number just offers to suppress IDE0059 and unrelated code fixes (whitespace for example).

To my knowledge, this foreach loop is the only way to count the number of `ReadOnlyMemory<byte>` that make up the `ReadOnlySequence<byte>`, the `Length` property only counts the total number of bytes in the entire sequence.

### LINKED ISSUE #32924: "Remove unused parameter" IDE0060 triggered for captured parameter
**Version Used**: 
16.0.0 Preview 2.0

**Steps to Reproduce**:
```csharp
public void Test(int foo) // IDE0060 triggered for foo
    => Console.CancelKeyPress += (s, e) => e.Cancel = foo != 0;
```

**Expected Behavior**:
`foo` is used, so no warning.

**Actual Behavior**:
The warning is shown, there's a code fix light bulb when hovering over the parameter, but that one disappears when being clicked.

### LINKED ISSUE #32946: IDE0059 - Remove redundant assignment triggered for variable used from closure
**Version Used**: VS 16.0 Preview 2.0

**Steps to Reproduce**:

```csharp
int i = 1;
Action a = () => Console.WriteLine(i);
```

**Expected Behavior**:

IDE0059 is not fired, and no suggested fix to remove a redundant assignment is made.

**Actual Behavior**:

IDE0059 fires and suggests to remove the redundant assignment, which results in:

```csharp
int i;
Action a = () => Console.WriteLine(i);
```

which doesn't compile. Adding a use site outside the lambda/delegate body makes IDE0059 go away.

**Context**:

The bigger repro for this issue is a place where we cache delegates that have a reference to a loop variable, which is copied to a local inside the body of a `for` loop, like this:

```csharp
Action[] GetCachedDelegates(int n)
{
    var res = new Action[n];

    for (var i = 0; i < n; i++)
    {
        var j = i;
        res[i] = () => DoSomething(j); // closure over `this` and `j`
    }

    return res;
}
```

### LINKED ISSUE #32959: Unused Parameter analysis is not resilient to bad invocations
```C#
using System;

namespace ConsoleApp1
{
    class Program
    {
        static void Main()
        {
            Test(true, false);
        }

        static void Test(bool arg1, bool arg2)
        {
            Invoke<string>(() => arg1 && arg2);

            T Invoke<T>(Func<T> a) { return a(); }
        }
    }
}
```

`arg1` and `arg2` are grayed out in the ide for being unused. Removing the `<string>` constraint causes all errors in the code to go away, and the parameters are correctly inferred to be used.

### LINKED ISSUE #32973: IDE0060 false positive (unused parameter)
**Version Used**:  `16.0.0 Preview 3.0 [28529.153.d16.0]`

**Steps to Reproduce**:

```csharp
class C
{
    public static bool M(out int x)
    {
        return LocalFunction(out x);

        bool LocalFunction(out int y)
        {
            y = 0;
            return true;
        }
    }
}
```

**Expected Behavior**:
No IDE0060

**Actual Behavior**:
```
Severity	Code	Description	Project	File	Line	Suppression State
Message	IDE0060	Remove unused parameter 'y'	ConsoleApp6	C:\Users\mavasani\source\repos\ConsoleApp6\ConsoleApp6\Program.cs	7	Active
```

### PR DESCRIPTION
**Pull request by mavasani**
FYI: I would recommend reviewing one commit at a time.

1. https://github.com/dotnet/roslyn/commit/e25cc1428ad511e24ed1c8b6905b0879226d356d: Bail out from unused value assignment reporting for special symbols names (`_`, `_1`, `_2`, etc.)
Fixes #32923 

2. https://github.com/dotnet/roslyn/commit/1ea6505fc74c1262d6bb897603ade00d3ef89020: Bail out from flow analysis in presence of invalid operations.
Fixes #32959

3. https://github.com/dotnet/roslyn/commit/59350e71d9fe0a0460ceae409d85b0a5b89e7826: Fix the logic that decides reporting unused ref/out parameters of local functions.
Fixes #32973

4. https://github.com/dotnet/roslyn/pull/33036/commits/ba127dc792c66462003a15218e2f4dd5a3412383: Make the analyzer conservative for handled delegate creation cases. Prior implementation looked for specific unhandled operation tree shapes and was agressive, leading to false positives. New implementation handles only specific operation tree shapes and is much more conservative.
Fixes #32946
Fixes #32924

### CODE REVIEW

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs [hunk 1]
```diff
@@ -448,9 +448,11 @@ private void AnalyzeOperationBlockEnd(OperationBlockAnalysisContext context)
                         //   2. Symbol has error type, hence the diagnostic could be noised
                         //   3. Static local symbols. Assignment to static locals
                         //      is not unnecessary as the assigned value can be used on the next invocation.
+                        //   4. Ignore special discard symbol names (see https://github.com/dotnet/roslyn/issues/32923).
                         if (_options.UnusedValueAssignmentSeverity == ReportDiagnostic.Suppress ||
                             symbol.GetSymbolType().IsErrorType() ||
-                            symbol.IsStatic && symbol.Kind == SymbolKind.Local)
+                            symbol.IsStatic && symbol.Kind == SymbolKind.Local ||
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs]: i hate the mix of && and ||.  can you parenthesize to make precedence clear?

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs [hunk 2]
```diff
@@ -312,6 +319,14 @@ private bool ShouldAnalyze(IOperation operationBlock, ISymbol owningSymbol)
...
 
+                    //  7. Bail out on invalid operations, i.e. code with semantic errors.
+                    //     We are likely to have false positives from flow analysis results
+                    //     as we will not account for potential lambda/local function invocations.
+                    if (_hasInvalidOperation)
+                    {
+                        return false;
+                    }
+
                     //  7. Otherwise, we execute analysis by walking the reaching symbol write chain to attempt to
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs]: need to renumber.

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs [hunk 3]
```diff
@@ -426,7 +426,30 @@ private void AnalyzeOperationBlockEnd(OperationBlockAnalysisContext context)
...
                                 {
                                     var hasReference = symbolUsageResult.SymbolsRead.Contains(unusedParameter);
-                                    _symbolStartAnalyzer.ReportUnusedParameterDiagnostic(unusedParameter, hasReference, context.ReportDiagnostic, context.Options, context.CancellationToken);
+
+                                    bool shouldReport;
+                                    switch (unusedParameter.RefKind)
+                                    {
+                                        case RefKind.Out:
+                                            // Do not report out parameters of local functions.
+                                            shouldReport = false;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs]: This needs explanation as to why.

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs [hunk 4]
```diff
@@ -195,55 +178,110 @@ private void AnalyzeExpressionStatement(OperationAnalysisContext context)
...
+                    return;
+
+                    // Local functions.
+                    bool IsHandledOperationTreeShape(IOperation operation)
                     {
-                        _hasConversionFromDelegateTypeOrAnonymousFunctionToNonDelegateType = true;
+                        // We only handle certain operation tree shapes in flow analysis of delegate creations
+                        // and consider those as non-delegate escaping operation trees.
+                        // For the remaining unknown ones, we conservatively mark the operation to lead to
+                        // delegate escape, and corresponding bail out from flow analysis in ShouldAnalyze method below.
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs]: thanks for the comment.

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs [hunk 5]
```diff
@@ -195,55 +178,110 @@ private void AnalyzeExpressionStatement(OperationAnalysisContext context)
...
+
+                        // 1. Delegate creation or anonymous function variable initializer is handled.
+                        if (operation.Parent is IVariableInitializerOperation)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs]: i'd prefer examples with explanation/intution.

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs [hunk 6]
```diff
@@ -195,55 +178,110 @@ private void AnalyzeExpressionStatement(OperationAnalysisContext context)
...
-                }
 
-                private void AnalyzeParameterReference(OperationAnalysisContext operationAnalysisContextContext)
-                {
-                    var parameter = ((IParameterReferenceOperation)operationAnalysisContextContext.Operation).Parameter;
-                    _referencedParameters.GetOrAdd(parameter, true);
+                    return;
+
+                    // Local functions
+                    bool IsHandledOperationTreeShape(IOperation operation)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs]: can we explan why we need both of these local functions (that are called the same thing)?  and seem to be checking the same thing?  Just a small intuition about the breakdown of why we need both, and which cases go with wwhich.

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs (hunk 2 above)
Comment (mavasani) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs]: The numbering got corrected in subsequent commits, verified current numbering is fine.

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs (hunk 5 above)
Comment (mavasani) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs]: Done, hope it is more clear.

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs (hunk 6 above)
Comment (mavasani) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs]: First local function is checking operation tree shape in context of a delegate creation or anonymous function, while the second one is check the tree shape in presence of a parameter/local reference of a delegate type.

Comment (mavasani) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs]: Refactored, renamed and added doc comments to methods. Hopefully it is more clear now.

**Comment by mavasani** [comment]
@genlu For another set of eyes on the PR.

@jinujoseph @vatsalyaagrawal for approval.

**Comment by genlu** [comment]
So for "delegate escaping" analysis, we switched from a blacklist to whitelist? I like it.

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs [hunk 7]
```diff
@@ -188,55 +178,141 @@ private void AnalyzeExpressionStatement(OperationAnalysisContext context)
...
+                    //    and analyze lambda body at invocation sites 'a();' 
+                    if (operation.Parent is ISimpleAssignmentOperation assignment &&
+                        (assignment.Target.Kind == OperationKind.LocalReference ||
+                         assignment.Target.Kind == OperationKind.ParameterReference))
+                    {
+                        return true;
+                    }
+
+                    // 3. For anonymous functions parented by delegate creation, we analyze the parent operation.
+                    //    For example, for 'Action a = () => { ... };', the lambda generates an anonymous function
```
[the remark this answers is not in the document]
Comment (genlu) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs]: If this is the case, consider add comment/assert here.

**Pull request merged**